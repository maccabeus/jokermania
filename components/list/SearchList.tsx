import { FC, Suspense, useEffect, useState } from "react";
import { ErrorBoundary } from "react-error-boundary";

import { JokeItemProps, SearchListProps } from "../types";
import JokeCard from "../card/JokeCard";
import ErrorCard from '../card/ErrorCard';
import LoadingCard from '../card/LoadingCard';
import { ApiCallStateEnum, IApiGetData } from "../../services/types";
import { BaseList } from "./BaseList";

const SearchList: FC<SearchListProps> = (props) => {

    const { searchResult } = props;

    const JokeRenderer = (joke: JokeItemProps) => {
        return <JokeCard item={joke.item} index={joke.index} />
    }
    const keyExtractor = (joke: JokeItemProps) => {
        return joke.id;
    }

    const RenderSearchList = (props: { searchResult: IApiGetData }) => {

        if (!searchResult || !searchResult.data) {
            return null;
        }
        const { data, status } = searchResult;
        let searchData: Array<JokeItemProps> = data ?? [];

        if (status === ApiCallStateEnum.ERROR) {
            return <ErrorCard title="Error occurs while  searching" />
        }
        if (status === ApiCallStateEnum.LOADING) {
            return <LoadingCard title={`Searching...`} />
        }
        if (status === ApiCallStateEnum.COMPLETED && searchData.length <= 0) {
            return <LoadingCard title={`No result found`} />
        }

        return (
            <BaseList
                data={searchData}
                renderItem={JokeRenderer}
                keyExtractor={keyExtractor}
            />
        )
    }
    return (
        <ErrorBoundary fallback={<ErrorCard />}>
            <Suspense fallback={<LoadingCard />}>
                <RenderSearchList searchResult={searchResult} />
            </Suspense>
        </ErrorBoundary>

    )
}

export default SearchList;
