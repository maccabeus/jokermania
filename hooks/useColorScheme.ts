import { ColorSchemeName, useColorScheme as _useColorScheme } from 'react-native';
import Layout from '../constants/Layout';
import { light, dark } from '../constants/Colors';

export type AppColorScheme = NonNullable<ColorSchemeName>;
export type AppColors = typeof light | typeof dark;
export type Layout = typeof Layout;

export type AppPropKeys = "theme" | any;

export type AppTheme = {
  color: AppColors;
  layout: Layout;
}

export type AppColorSchemeData = {
  [propName: string]: AppColors;
}

export type AppThemeContext = {
  [propName in AppPropKeys]: AppTheme
}
// export type AppThemeContext = {
//   theme: AppTheme
// }

export function useThemeSettings(colorScheme: AppColorScheme): AppTheme {
  return (
    {
      color: colorScheme === "dark" ? dark : light,
      layout: Layout
    }
  );
}

export default function useColorScheme(): AppColorScheme {
  return _useColorScheme() as AppColorScheme;
}