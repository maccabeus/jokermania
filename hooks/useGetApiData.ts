
import { callApi } from '../services/ApiService';
import { useState, useEffect } from 'react';
import { ApiCallStateEnum, IApiGetData } from '../services/types';

export type Predicate = ((...args: any) => boolean) | null;

export const useGetApiData = (apiPath: string, dataIndex: string | null = null, canFetch: Predicate = null): IApiGetData => {
    return fetchApiData(apiPath, null, dataIndex, canFetch);
}

export const fetchApiData = (apiPath: string, queryParam: string | null = null, dataIndex: string | null = null, canFetch: Predicate = null): IApiGetData => {

    const [status, setStatus] = useState(ApiCallStateEnum.IDLE);
    const [data, setData] = useState(undefined);
    const [success, setSuccess] = useState(false);

    useEffect(() => {
        if ((canFetch && canFetch() === true) || !canFetch) {
            /** set loading status */
            const apiFullPath = queryParam ? `${apiPath}=${queryParam}` : apiPath;
            setStatus(ApiCallStateEnum.LOADING);
            callApi(apiFullPath, "GET").then(response => {
                const { success, data } = response;
                if (success) {
                    // if there is a data index to retrieve, dig a bit deeper in  
                    dataIndex ? setData(data[dataIndex]) : setData(data);
                } else {
                    setData(undefined);
                }
                /** set the call success and status values */
                setSuccess(success);
                setStatus(ApiCallStateEnum.COMPLETED);
            }).catch((e: any) => {
                setStatus(ApiCallStateEnum.ERROR);
            })
        }
    }, [apiPath]);

    return { success, status, data, }
}