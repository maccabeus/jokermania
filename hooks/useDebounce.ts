import { Predicate } from "./useGetApiData";

const debounce = (waitTime: number = 1000, maxKeyStrokes: number = 3) => {

    let timerInterval: any = null;

    const debounceAction = (callback: Function, actionPredicate: Predicate) => {
        if (timerInterval) clearTimeout(timerInterval);

        if (actionPredicate && actionPredicate() === true) {
            timerInterval = setTimeout(() => {
                callback();
            }, waitTime);
        }
    }
    return debounceAction
}
const debounceOld = (waitTime: number = 2000, maxKeyStrokes: number = 3) => {

    let keyCount: number = 0;
    let timerInterval: any = null;

    const debounceAction = (callback: Function, actionPredicate: boolean = true) => {
        if (timerInterval) clearTimeout(timerInterval);
        keyCount += 1;
        console.log("key count", keyCount);
        if (actionPredicate) {
            timerInterval = setTimeout(() => {
                callback();
                /** reset keyCount */
                keyCount = 0;
                console.log("calling now count: ", keyCount);
            }, waitTime);
        }
    }
    return debounceAction
}

export default debounce;