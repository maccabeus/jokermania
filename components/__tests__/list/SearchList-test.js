import * as React from 'react';
import renderer from 'react-test-renderer';
import { ThemeProvider } from 'styled-components/native';
import { useThemeSettings } from '../../../hooks/useColorScheme';
import { jest } from '@jest/globals';
import SearchList from '../../list/SearchList';

describe("Search List Snapshot", () => {

    const scheme = jest.fn();
    let lightScheme = scheme.mockRejectedValue("light");
    let currentScheme = null;
    let searchResult = [];

    beforeAll(() => {
        currentScheme = useThemeSettings(lightScheme);
        searchResult = [
            { index: 0, item: { value: "First test joke" }, onClick: () => null },
            { index: 2, item: { value: "Second test joke" }, onClick: () => alert("ok") },
            { index: 3, item: { value: "Third test joke" }, onClick: () => undefined },
        ];
    });

    test(`Search List renders correctly`, () => {
        const tree = renderer.create(
            <ThemeProvider theme={currentScheme}>
                <SearchList searchResult={searchResult} />
            </ThemeProvider>
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });

    afterAll(() => {
        currentScheme = null;
        currentScheme = null;
    });
});


