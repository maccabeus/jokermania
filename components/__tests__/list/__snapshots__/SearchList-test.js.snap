// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search Button Snapshot Search List renders correctly 1`] = `null`;

exports[`Search List Snapshot Search List renders correctly 1`] = `null`;
